### ISSUE TITLE
LTFB produces garbage after first two LTFB rounds

### ISSUE THREAD

**Issue by timmoon10** [LLNL/lbann#1596]
Bug report from @samadejacobs:

> I write to inform you that PR1546 (mixed precision support) broke training with LTFB. I am putting together a unit test to catch this early. In the meantime, I have given you, Brian and Tim a reduced version of my experiment’s scripts (10K JAG dataset, which should in less than 5 minutes with 12 or 16 Lassen nodes). The problem as you will find out is that metrics and objective functions are “nan”.

When I run LTFB with AlexNet (1 Lassen node, 2 or 4 trainers, 1 proc/trainer), the first LTFB round proceeds correctly. During the second LTFB round, some of the trainers have garbage validation accuracy. I can't find a pattern in what trainers fail, although it appears deterministic. By the third LTFB round, all trainers produce garbage.

I can't think of an obvious reason why #1546 would cause this failure. The LTFB callback `dynamic_cast`s all of the weights to `data_type_weights<DataType>`:

https://github.com/LLNL/lbann/blob/2cb48bab4fceeb7fb4ec2b5054cf5fe52755dbaf/src/callbacks/ltfb.cpp#L471

If we're not doing any mixed precision stuff in the model, this should be well-behaved.

**Comment by timmoon10** [comment]
I think I've narrowed things down. Inside the LTFB callback, we call `data_type_weights`'s copy-assignment operator:

https://github.com/LLNL/lbann/blob/9502b2d6a050d4d9834ac638e87a639ee233a2c1/src/callbacks/ltfb.cpp#L523

The weight values are still sensible at this point. However, I see that `weights`'s copy-assignment operator does not copy `std::unordered_map<std::type_index, gradient_manager_ptr> gradients_`:

https://github.com/LLNL/lbann/blob/9502b2d6a050d4d9834ac638e87a639ee233a2c1/src/optimizers/optimizer.cpp#L61

During the next mini-batch step the values turn in to NaNs inside `data_type_optimizer::step()`:

https://github.com/LLNL/lbann/blob/9502b2d6a050d4d9834ac638e87a639ee233a2c1/src/optimizers/data_type_optimizer.cpp#L140

This is even when I set the learning rate to zero. I suspect we are not properly setting the gradient status, so the optimizer receives uninitialized gradients and the optimizer step becomes `0 * NaN -> NaN`. This also explains why some trainers survive an LTFB round: if the uninitialized gradients don't have any NaNs, then the model might still be a state where it can learn.
